flink-streaming-java/src/main/java/org/apache/flink/streaming/api/graph/StreamingJobGraphGenerator.java

```diff
@@ -348,6 +348,13 @@ private StreamConfig createJobVertex(
 					"Did you generate them before calling this method?");
 		}
 
+		if (!streamGraph.getExecutionConfig().hasAutoGeneratedUIDsEnabled()) {
```

Comment:
I don't think this is the best place to check it. 

Why don't we move the check to `org.apache.flink.streaming.api.graph.StreamGraphGenerator#transform`? I think what we want to effectively enforce is that a `StreamTransformation` has either `uid` or `userHash` set, don't we?

Reply:
Unfortunately we can’t do that, flink only uses the uid for the first operator in each operator chain. Most users seem to understand that and only set UIDs in those places. This is the only place we can check that the opeator id assigned to each job vertex is from a manually set uid which is what we’re really after.

Reply:
I tend to disagree. AFAIK for state mapping purposes all OperatorIds are preserved, even those that are chained into a single operator chain (a single chained JobVertex), otherwise it would be impossible to e.g. restore job with changed chaining, right?

Reply:
Only the uid for the head operator in a chain is used to generate the OperatorId for a JobVertex, and I am not familiar with uids being anywhere else to partition state within a single vertex.  I believe it is impossible to restore a job with changed chaining but I will double check.

Reply:
I am pretty sure you can, and all operators are used to generate OperatorId. See `org.apache.flink.streaming.api.graph.StreamingJobGraphGenerator#createJobVertex:368-394`. 

Also see the ctor of `JobVertex`, it's true that the id of head operator is used as `primaryId`, but all chained operator ids are passed as `operatorIds`, which is later used in e.g. `org.apache.flink.runtime.checkpoint.StateAssignmentOperation#assignTaskStateToExecutionJobVertices` to assign proper state.

Reply:
Thanks for pointing this out, I had no idea you could restore with different chaining. I'll make the necessary updates latter today.